Added module to get similarity in DB for custom input model/module

ShobhitBehl opened the pull request:
Added file to print top similar models/modules in database to an input file which is not present in the database. File is parsed using parser modules and fitted to Graph2Vec with the database models/modules.

ShobhitBehl committed:
Add common/Vectorize.py

Added common/Vectorize module to create graph embeddings using models in
database using Graph2Vec.
Only the geature 'operator_type' considered for a node.
Added process_nodes() functionality to common/Graph.py to compute
reachability of nodes from input and discard if unreachable.
Added a call to process_nodes() from TFLite/TFLiteParser.py.
Bug fix in TF/main.py to return on finding None graph in load_data().
Added is_input boolena column in Operators Table to identify input operators,
made changes in common/Storage.py to load the same.

ShobhitBehl committed:
Add CLA for edge and node attributes in feature

Added class attributes to define the attribute names from Node and Edge
to be used when building feature for graph2vec.
Added CLA to optionally use Edge and Ndode attributes listed in the class
attributes.
Added binary build for common/Vectorize.py.
Updated README.md

ShobhitBehl committed:
Add common/similarity.py

Added common/similarity.py to run similarity of embeddings, removed the
same from common/Vectorize.py for it to be strictly an embedding module.
Modified build for 'Vectorize' into a library and added binary build for
'similarity'.
Added "None" category support in load_data.py
Updated README.md

ShobhitBehl committed:
Add time performance measures

Added time performance measure to check time to query database models
and create Graphs and time to fit graph2vec model.
Fixed a bug with interchanged values of include_node_attrs and
include_edge_attrs.
Added docstring for Graph2Vec parameters.
Updated common/visualize.ipynb.

ShobhitBehl committed:
Add Embeddings to Models Table

Added embeddings storage in Models table in the database, added
functions for storing the same in common/Storage.py and
common/similarity.py.
Added wl_iterations as CLA.
Moved parse_models() functionality from common/Vectorize.py
to common/Storage.py.
Made signature changes to get_graph2vec_embeddings() in
common/Vectorize.py for it to take a list of model graphs as
argument instead of parsing models from database.

ShobhitBehl committed:
Add common/compare.py

Added common.compare.py, a module to find similarity in db models for a
custom input file.
Added bazel build for the same.
Fixed imports to assume base directory as ml-model-characterization.
Updated CLAs without default values as "required".
Updated README.md.

rachitagrawal commented on common/compare.py:
```diff
@@ -0,0 +1,111 @@
+        parser = TFParser.TFParser()
+        graph = parser.parse_graph(file_path, model_name, None, None, "True", [])
+
+    else:
```
Here if the file_format is neither "TFFrozenGraph" nor "TFSavedModel", we assume it would be "TFLiteModel". What if user doesn't give any of these three? Can we throw proper error message in such case?

ShobhitBehl replied:
Added an error message. Same is also handled by 'choices' field in argparse.

rachitagrawal commented on common/compare.py:
```diff
@@ -0,0 +1,111 @@
+
+    print(similarity)
+    
+    model_graph = model_graphs[len(model_graphs) - 1]
```
Is it always guaranteed that get_graph2vec_embedding will not return empty 'model_graphs'?

ShobhitBehl replied:
It is guaranteed that 'model_graphs' will contain at least the file being parsed, added a check for failed parsing.

rachitagrawal commented on common/compare.py:
```diff
@@ -0,0 +1,111 @@
+
+    # Cosine similarity
+    similarity = metrics.pairwise.cosine_similarity(
+        [embeddings[len(embeddings) - 1]], embeddings)
```
Is it always guaranteed that get_graph2vec_embedding will not return empty 'embeddings'?

ShobhitBehl replied:
Guaranteed that 'embeddings' will not be empty since 'model_graphs' will contain at least the file being parsed.
Even in other cases, empty 'embeddings' only possible on empty 'model_graphs' in which case graph2vec will throw an error.

rachitagrawal commented on common/compare.py:
```diff
@@ -0,0 +1,111 @@
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--file_path', required = True)
+    parser.add_argument('--file_format', required = True)
```
You can use 'choices' field to control the allowed values.

ShobhitBehl replied:
Added 'choices' for argparse in file_format.

rachitagrawal commented on common/compare.py:
```diff
@@ -0,0 +1,111 @@
+
+    file_path = args.file_path
+    file_format = args.file_format
+    model_name = args.model_name
```
What if the given model_name exists in db? Does it have any unexpected behavior in that case?

ShobhitBehl replied:
Code runs without any unexpected behavior, prints similar models as expected.

ShobhitBehl committed:
Add 'choices' in argparse for file_format

Added 'choices' in argparse for file_format for control over input from
user.
Added error message in case of function being used with wrong
file_format.

Fix operator name AveragePool to AvgPool

Fixed a bug where operator name for mapped wrongly to AveragePool
instead of AvgPool.

Update common/visualize.ipynb

rachitagrawal reviewed (approved):
LGTM

ShobhitBehl committed:
Update common/visualize.ipynb

Update common/visualize.ipynb to resolve merge conflict.
